kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala

```diff
@@ -233,48 +288,148 @@ class KinesisSink private (
         }
       }
     }
 
-  private def multiPut(name: String, batch: List[(ByteBuffer, String)]): Future[PutRecordsResult] =
+  private def sendToSqsOrRetryToKinesis(
+    failures: List[BatchEntry],
+    nextBackoff: Long
+  )(
+    retryErrorMsg: String
+  ): Unit =
+    maybeSqs match {
+      case Some(sqs) =>
+        log.info(
+          s"Sending ${failures.size} events from a batch to SQS buffer queue: ${sqs.sqsBufferName}"
+        )
+        putToSqs(sqs, failures)
+        ()
+      case None =>
+        log.error(retryErrorMsg)
+        log.warn(
+          s"${failures.size} failed events scheduled for retry (consider setting a SQS Buffer in config.hocon)"
+        )
+        scheduleBatch(failures, nextBackoff)
+    }
+
+  private def putToSqs(sqs: SqsClientAndName, batch: List[BatchEntry]): Future[Unit] =
+    Future {
+      log.info(s"Writing ${batch.size} messages to SQS queue: ${sqs.sqsBufferName}")
+      val sqsBatchEntries = batch.map(addUuids).map(toSqsBatchEntry)
+      val MaxSqsBatchSize = 10
+      sqsBatchEntries.grouped(MaxSqsBatchSize).foreach { batchEntryGroup =>
+        sendToSqsWithRetry(sqs, batchEntryGroup)
+      }
+    }
+
+  /**
+   *  UUIDs are generated for retry purpose. When batch fails, we want to retry (limited number of times)
+   *   only those entries that failed within the batch.
+   */
+  private val addUuids: (BatchEntry) => (BatchEntryWithId) = {
+    case BatchEntry(msg, key) => BatchEntryWithId(msg, key, java.util.UUID.randomUUID())
+  }
+
+  private val toSqsBatchEntry: (BatchEntryWithId) => (UUID, SendMessageBatchRequestEntry) = {
+    case BatchEntryWithId(msg, key, id) =>
+      val b64EncodedMsg = encode(msg)
+      val batchReqEntry = new SendMessageBatchRequestEntry(UUID.randomUUID.toString, b64EncodedMsg)
+        .withMessageAttributes(
+          Map(
+            "kinesisKey" ->
+              new MessageAttributeValue()
+                .withDataType("String")
+                .withStringValue(key)
+          ).asJava
+        )
+        .withId(id.toString)
+      (id, batchReqEntry)
+  }
+
+  private def createBatchRequest(queueUrl: String, batch: List[SendMessageBatchRequestEntry]) =
+    new SendMessageBatchRequest()
+      .withQueueUrl(queueUrl)
+      .withEntries(batch.asJava)
+
+  private def sendToSqsWithRetry(
+    sqs: SqsClientAndName,
+    batchEntryGroup: List[(UUID, SendMessageBatchRequestEntry)]
+  ) = {
+    val current = TrieMap(batchEntryGroup: _*)
+    retry
+      .JitterBackoff(10, 1.second)
+      .apply { () =>
+        Future {
+          val batchRequest = createBatchRequest(sqs.sqsBufferName, current.values.toList)
+          val res = sqs.sqsClient.sendMessageBatch(batchRequest)
+          val failed = res.getFailed().asScala
+          if (failed.nonEmpty) {
+            val errors = failed.map(_.toString).mkString(", ")
+            log.error(s"Sending to SQS queue: ${sqs.sqsBufferName} failed with: $errors")
```

Comment:
```suggestion
            log.error(s"Sending to SQS queue [${sqs.sqsBufferName}] failed with errors [$errors]")
```
We usually use brackets to clearly indicate what comes from a variable

Reply:
Really? Sounds weird, brackets indicates tables usually.